flink-ml-core/src/main/java/org/apache/flink/ml/linalg/BLAS.java

```diff
@@ -32,9 +32,31 @@ public static double asum(DenseVector x) {
     }
 
     /** y += a * x . */
-    public static void axpy(double a, DenseVector x, DenseVector y) {
+    public static void axpy(double a, Vector x, DenseVector y) {
         Preconditions.checkArgument(x.size() == y.size(), "Vector size mismatched.");
-        JAVA_BLAS.daxpy(x.size(), a, x.values, 1, y.values, 1);
+        if (x instanceof SparseVector) {
+            axpy(a, (SparseVector) x, y);
+        } else {
+            axpy(a, (DenseVector) x, y);
+        }
+    }
+
+    /** Computes the hadamard product of the two vectors (y = y \hdot x). */
+    public static void hDot(Vector x, Vector y) {
+        Preconditions.checkArgument(x.size() == y.size(), "Vector size mismatched.");
+        if (y instanceof DenseVector) {
+            if (x instanceof SparseVector) {
```

Comment:
Thanks for the comments. I have reordered the order of `a instanceof b`.

BTW, we do not support `axpy` when `y` is a SparseVector because it may lead to some memory re-allocation.